pulsar-io/jdbc/src/main/java/org/apache/pulsar/io/jdbc/JdbcAbstractSink.java

```diff
@@ -0,0 +1,187 @@
+    // ----- Runtime fields
+    private JdbcSinkConfig jdbcSinkConfig;
+    @Getter
+    private Connection connection;
+    private String tableName;
+    private String schema;
+
+    private JdbcUtils.TableId tableId;
+    private JdbcUtils.TableDefinition tableDefinition;
+    private PreparedStatement insertStatement;
+
+    // for flush
+    private List<Record<T>> incomingList;
+    private List<Record<T>> swapList;
+    private AtomicBoolean isFlushing;
+    private int timeoutMs;
+    private int batchSize;
+    private ScheduledExecutorService flushExecutor;
+
+    @Override
+    public void open(Map<String, Object> config, SinkContext sinkContext) throws Exception {
+        // TODO: currently assume schema is same as created table schema;
+        // turn to getSchema from SinkContext.getTopicSchema.getSchema(inputTopic)?
+
+        jdbcSinkConfig = JdbcSinkConfig.load(config);
+
+        String jdbcUrl = jdbcSinkConfig.getJdbcUrl();
+        if (jdbcSinkConfig.getJdbcUrl() == null) {
+            throw new IllegalArgumentException("Required jdbc Url not set.");
+        }
+
+        Properties properties = new Properties();
+        String username = jdbcSinkConfig.getUserName();
+        String password = jdbcSinkConfig.getPassword();
+        if (username != null) {
+            properties.setProperty("user", username);
+        }
+        if (password != null) {
+            properties.setProperty("password", password);
+        }
+
+        connection = JdbcUtils.getConnection(jdbcUrl, properties);
+        log.info("Connection open");
+
+        schema = jdbcSinkConfig.getSchema();
+        tableName = jdbcSinkConfig.getTableName();
+        tableId = JdbcUtils.getTableId(connection, tableName);
+        tableDefinition = JdbcUtils.getTableDefinition(connection, tableId);
+        insertStatement = JdbcUtils.buildInsertStatement(connection, JdbcUtils.buildInsertSql(tableDefinition));
+
+        timeoutMs = jdbcSinkConfig.getTimeoutMs();
+        batchSize = jdbcSinkConfig.getBatchSize();
+        incomingList = Lists.newArrayList();
+        swapList = Lists.newArrayList();
+        isFlushing = new AtomicBoolean(false);
+
+        flushExecutor = Executors.newScheduledThreadPool(1);
+        flushExecutor.scheduleAtFixedRate(() -> flush(), timeoutMs, timeoutMs, TimeUnit.MILLISECONDS);
+    }
+
+    @Override
+    public void close() throws Exception {
+        if (!connection.getAutoCommit()) {
+            connection.commit();
+        }
+        flushExecutor.shutdown();
+        if (connection != null) {
+            connection.close();
+        }
+        log.info("Connection Closed");
+    }
+
+    @Override
+    public void write(Record<T> record) throws Exception{
+        synchronized (incomingList) {
+            incomingList.add(record);
+        }
+
+        if (incomingList.size() >= batchSize) {
+            flushExecutor.schedule(() -> flush(), 0, TimeUnit.MILLISECONDS);
```

Comment:
you can scheduler a periodically flush when you open the sink.

Reply:
Thanks, Yes, there was already one. The thought it, besides a periodically flush, also have a size to controll the flush.